library/std/src/backtrace.rs

```diff
@@ -346,6 +347,19 @@ impl Backtrace {
             Inner::Captured(_) => BacktraceStatus::Captured,
         }
     }
+
+    /// Returns an iterator over the backtrace frames.
+    #[unstable(feature = "backtrace_frames")]
+    pub fn frames(&self) -> &[BacktraceFrame] {
+        let frames = match self.inner {
+            Inner::Captured(c) => {
+               let captured = c.lock().unwrap();
```

Comment:
This might be a bit of an issue for us. We might end up needing to wrap this up in something like:

```rust
pub struct Frames<'a> {
    inner: Option<MutexGuard<'a, Capture>>,
}

impl<'a> AsRef<[BacktraceFrame]> for Frames<'a> {
    fn as_ref(&self) -> &[BacktraceFrame] {
        match self.inner {
            Some(captured) => &*captured.frames,
            None => &[],
        }
    }
}
```

Reply:
The reason being the lifetime of the frames we can borrow from that locked mutex won't be enough for us to return a reference to the slice of frames.

Reply:
Thinking about this more, we probably don't want to hand out a quiet borrow to locked state, otherwise we might create the potential for deadlocks. Instead, we could come up with something like this:

```rust
pub struct Frames {
    inner: Vec<BacktraceFrame>,
}

impl<'a> AsRef<[BacktraceFrame]> for Frames<'a> {
    fn as_ref(&self) -> &[BacktraceFrame] {
        &self.inner
    }
}
```

And materialize that `Vec` under the lock.

Reply:
I'm not quite sure what you mean by "materialize that `Vec` under the lock". Is that just referring to the `Frames` type holding on to a `Vec<BacktraceFrame>` so that we don't need to manually deal with the lock?

Reply:
Ah sorry, yes I meant making our `frames()` method look something like this:

```rust
impl Backtrace {
    pub fn frames(&self) -> Frames {
        if let Inner(captured) = self.inner {
            Frames {
                frames: captured.lock().unwrap().frames.clone(),
            }
        } else {
            Frames {
                frames: vec![],
            }
        }
    }
}
```

So that we're not handing out a borrow from that lock

Reply:
I think `Frames` should have a lifetime tied to `Backtrace` even if it currently isn't used. In the future a pure rust unwinder may avoid the lock. We could then generate the frames on the fly in the `Iterator` impl for `Frames`, avoiding a memory allocation.